File: framework/languages/java/e2e/InstanceSpot/Owner/src/main/java/systems/zlink/e2e/instancespot/owner/EvidenceStore.java
package systems.zlink.e2e.instancespot.owner;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicLong;
import systems.zlink.e2e.instancespot.shared.Contracts;
import systems.zlink.e2e.instancespot.shared.Wait;

public final class EvidenceStore {
    private final OwnerOptions options;
    private final AtomicLong nextSequence = new AtomicLong();
    private final List<Contracts.EvidenceEvent> events = new ArrayList<>();

    public EvidenceStore(OwnerOptions options) {
        this.options = options;
    }

    public String rid() {
        return options.rid();
    }

    public String lifecycleId() {
        return options.lifecycleId();
    }

    public synchronized Contracts.EvidenceEvent record(
        String kind,
        String spotId,
        String operationId,
        String payload,
        long generation,
        int activeHandlers,
        String detail) {
        Contracts.EvidenceEvent event = new Contracts.EvidenceEvent(
            nextSequence.incrementAndGet(),
            kind,
            spotId == null ? "" : spotId,
            operationId == null ? "" : operationId,
            payload == null ? "" : payload,
            options.rid(),
            options.lifecycleId(),
            generation,
            activeHandlers,
            detail == null ? "" : detail);
        events.add(event);
        append(event);
        return event;
    }

    public synchronized Contracts.EvidenceSnapshot snapshot() {
        return new Contracts.EvidenceSnapshot(
            options.rid(), options.lifecycleId(), List.copyOf(events));
    }

    public Contracts.EvidenceWaitResult waitFor(
        Contracts.EvidenceWaitRequest request) {
        Contracts.EvidenceSnapshot result = Wait.until(
            Duration.ofMillis(request.timeoutMilliseconds()),
            "timed out waiting for Instance Spot evidence kind=" + request.kind()
                + " operation=" + request.operationId(),
            () -> {
                Contracts.EvidenceSnapshot current = snapshot();
                boolean found = current.events().stream().anyMatch(event ->
                    (request.kind().isBlank() || event.kind().equals(request.kind()))
                        && (request.operationId().isBlank()
                            || event.operationId().equals(request.operationId())));
                return found ? current : null;
            });
        return new Contracts.EvidenceWaitResult(true, result);
    }

    private void append(Contracts.EvidenceEvent event) {
        if (options.evidenceFile().isBlank()) {
            return;
        }
        try {
            Path path = Path.of(options.evidenceFile());
            Path parent = path.getParent();
            if (parent != null) {
                Files.createDirectories(parent);
            }
            Files.writeString(
                path,
                event.sequence() + "|" + event.kind() + "|"
                    + event.spotId() + "|" + event.operationId() + "|"
                    + event.payload() + "|" + event.objectGeneration() + "|"
                    + event.activeHandlers() + System.lineSeparator(),
                StandardCharsets.UTF_8,
                java.nio.file.StandardOpenOption.CREATE,
                java.nio.file.StandardOpenOption.APPEND);
        } catch (IOException error) {
            throw new IllegalStateException("failed to write Instance Spot evidence", error);
        }
    }
}

File: framework/languages/java/e2e-kotlin/InstanceSpot/Owner/src/main/kotlin/systems/zlink/e2e/kotlin/instancespot/owner/EvidenceStore.kt
package systems.zlink.e2e.kotlin.instancespot.owner

import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardOpenOption
import java.time.Duration
import java.util.concurrent.atomic.AtomicLong
import systems.zlink.e2e.kotlin.instancespot.shared.Contracts
import systems.zlink.e2e.kotlin.instancespot.shared.Wait

class EvidenceStore(private val options: OwnerOptions) {
    private val sequence = AtomicLong()
    private val events = mutableListOf<Contracts.EvidenceEntry>()

    @Synchronized
    fun record(
        kind: String,
        spotId: String = "",
        operationId: String = "",
        payload: String = "",
        generation: Long = 0,
        activeHandlers: Int = 0,
        detail: String = "",
    ): Contracts.EvidenceEntry {
        val event = Contracts.EvidenceEntry(
            sequence.incrementAndGet(), kind, spotId, operationId, payload,
            options.rid, options.lifecycleId, generation, activeHandlers, detail,
        )
        events += event
        append(event)
        return event
    }

    @Synchronized
    fun snapshot(): Contracts.EvidenceSnapshot =
        Contracts.EvidenceSnapshot(options.rid, options.lifecycleId, events.toList())

    fun waitFor(request: Contracts.EvidenceWaitReq): Contracts.EvidenceWaitRes {
        val snapshot = Wait.until(
            Duration.ofMillis(request.timeoutMilliseconds),
            "timed out waiting for Instance Spot evidence kind=${request.kind} operation=${request.operationId}",
        ) {
            snapshot().takeIf { current ->
                current.events.any { event ->
                    (request.kind.isBlank() || event.kind == request.kind) &&
                        (request.operationId.isBlank() || event.operationId == request.operationId)
                }
            }
        }
        return Contracts.EvidenceWaitRes(true, snapshot)
    }

    private fun append(event: Contracts.EvidenceEntry) {
        if (options.evidenceFile.isBlank()) return
        val path = Path.of(options.evidenceFile)
        path.parent?.let(Files::createDirectories)
        val line = listOf(
            event.sequence, event.kind, event.spotId, event.operationId,
            event.payload, event.objectGeneration, event.activeHandlers,
        ).joinToString("|") + System.lineSeparator()
        Files.writeString(
            path, line, StandardCharsets.UTF_8,
            StandardOpenOption.CREATE, StandardOpenOption.APPEND,
        )
    }
}
